ufl/algorithms/signature.py

```diff
@@ -94,19 +94,20 @@ def compute_terminal_hashdata(expressions, renumbering):
 
 
 def compute_expression_hashdata(expression, terminal_hashdata):
-    # The hashdata computed here can be interpreted as prefix operator
-    # notation, i.e. we store the equivalent of '+ * a b * c d' for
-    # the expression (a*b)+(c*d)
     expression_hashdata = []
-    for expr in pre_traversal(expression):
+
+    for expr in unique_pre_traversal(expression):
+        # Visit each node only once, but store all its operands
         if expr._ufl_is_terminal_:
-            data = terminal_hashdata[expr]
+            expression_hashdata.append(terminal_hashdata[expr])
         else:
-            data = expr._ufl_typecode_  # TODO: Use expr._ufl_signature_data_()? More extensible, but more overhead.
-        expression_hashdata.append(data)
-    # Oneliner: TODO: Benchmark, maybe use a generator?
-    # expression_hashdata = [(terminal_hashdata[expr] if expr._ufl_is_terminal_ else expr._ufl_typecode_)
-    #                       for expr in pre_traversal(expression)]
+            expression_hashdata.append(expr._ufl_typecode_)
+            for op in expr.ufl_operands:
```

Comment:
Why do you need to expand the first level of children for the signature? I would have thought that the unique pre-order visit is unique for signature purposes even without expanding the tree like this.

That is, why is this not a sufficient implementation?

```
return tuple(terminal_hashdata[expr] if expr._ufl_is_terminal_ else expr._ufl_typecode_ for expr in unique_pre_traversal(expression))
```
?

Reply:
I did it to avoid the following
```python
import ufl
from ufl.corealg.traversal import unique_pre_traversal

c1 = ufl.Constant(ufl.triangle)
c2 = ufl.Constant(ufl.triangle)

expression = c2*c2 + c1
print(tuple(str(expr) if expr._ufl_is_terminal_ else expr._ufl_typecode_ for expr in unique_pre_traversal(expression)))

expression = c2*c1 + c1
print(tuple(str(expr) if expr._ufl_is_terminal_ else expr._ufl_typecode_ for expr in unique_pre_traversal(expression)))

```
which produces the same hashdata
```
(61, 62, 'c_1', 'c_0')
(61, 62, 'c_1', 'c_0')
```

However, when I think about it, there could be rare cases where even first level children expansion produces same expr hashdata. Any suggestions? We need to avoid non-unique traversal.

Reply:
Could one do the following instead (I don't have the benchmarking setup to know if this is faster):

```python
class Hashdata(MultiFunction):
    def __init__(self, terminal_hashdata):
        super().__init__()
        self.terminal_hashdata = terminal_hashdata

    def expr(self, op, *ops):
        if op._ufl_is_terminal_:
            return (self.terminal_hashdata[op],)
        return tuple(op._ufl_typecode_, *ops)

def compute_expression_hashdata(expr, terminal_hashdata):
    mapper = Hashdata(terminal_hashdata)
    return map_expr_dag(mapper, expr)
```

Reply:
It is worse, `MultiFunction` is applied with `map_expr_dag` which uses `unique_post_traversal`, but the overhead here I think comes from all the caching and dictionary access (and possibly hashing the elements too?) - see the profile

<img width="1179" alt="Screenshot 2022-03-30 at 14 59 18" src="https://user-images.githubusercontent.com/2896171/160839927-b63f2f13-a7df-4070-ba21-4c089d521c91.png">
.

Reply:
Almost all of the cost looks to be hashing `FiniteElement` objects?

Reply:
Probably due to `rcache` in `map_expr_dags` and `FiniteElementBase` hash not being cached.

I've managed to achieve ~4s with unique post traversal and insertion of operands based on cache (similar to map_expr_dags, but simpler). Almost all time spent is now in `str(that_huge_expressiondata_list)` in `signature.py` line 161.

The form I am using is a jacobian computed with UFL AD from a potential, where the potential is defined using closed-form expressions for eigenvalues of 3x3 Cauchy-Green tensor. We are essentially differentiating through roots of characteristic polynomial of ugly symbolic matrix.

Reply:
I mean that's better, but still pretty bad. I wonder if it is faster to build the bytes array directly while traversing. `str` isn't taking advantage of the shared sub-structure so I imagine that could be more efficient.

Reply:
I am now applying hash at each node and storing hashed byte-strings for the `cache`, this seems to work very well (negligible perf cost).

Reply:
I am happy with this, any comments @wence- ?

Reply:
good to go now!